atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/CassandraImitatingConsensusForgettingStore.java

```diff
@@ -0,0 +1,279 @@
+                cell,
+                getQuorumNodes(),
+                node -> node.put(
+                        cell, ImmutableBytesAndTimestamp.of(value, KvsConsensusForgettingStore.PUT_TIMESTAMP)));
+    }
+
+    @Override
+    public void put(Map<Cell, byte[]> values) {
+        // sort by cells to avoid deadlock
+        KeyedStream.ofEntries(values.entrySet().stream().sorted(Map.Entry.comparingByKey()))
+                .forEach(this::put);
+    }
+
+    public void setProbabilityOfFailure(double newProbability) {
+        probabilityOfFailure = newProbability;
+    }
+
+    public Optional<BytesAndTimestamp> getInternal(Cell cell, Set<Node> quorumNodes) {
+        Set<Optional<BytesAndTimestamp>> reads = quorumNodes.stream()
+                .map(node -> Optional.ofNullable(node.get(cell)))
+                .collect(Collectors.toSet());
+        Optional<BytesAndTimestamp> result = reads.stream()
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .max(Comparator.comparing(BytesAndTimestamp::timestamp));
+        if (reads.size() > 1) {
+            runStateMutatingTaskOnNodes(cell, quorumNodes, node -> node.tryUpdateTo(cell, result.get()));
+        }
+        return result;
+    }
+
+    private void maybeFail() {
+        if (random.nextDouble() <= probabilityOfFailure) {
+            throw new SafeRuntimeException("Ohno!");
+        }
+    }
+
+    private void writeToQuorum(Cell cell, Set<Node> quorumNodes, byte[] value) {
+        ImmutableBytesAndTimestamp tsValue = ImmutableBytesAndTimestamp.of(value, timestamps.getAndIncrement());
+        runTaskOnNodesMaybeFail(quorumNodes, node -> {
+            node.put(cell, tsValue);
+        });
+    }
+
+    private Set<Node> getQuorumNodes() {
+        List<Integer> indices = IntStream.range(0, NUM_NODES).boxed().collect(Collectors.toList());
+        Collections.shuffle(indices);
+        return IntStream.range(0, QUORUM).mapToObj(indices::get).map(nodes::get).collect(Collectors.toSet());
+    }
+
+    private void runAtomically(Cell cell, Runnable task) {
+        ReentrantReadWriteLock.WriteLock lock =
+                getReentrantReadWriteLockForCell(cell).writeLock();
+        lock.lock();
+        try {
+            task.run();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private void runStateMutatingTaskOnNodes(Cell cell, Set<Node> quorumNodes, Consumer<Node> task) {
+        ReentrantReadWriteLock.ReadLock lock =
+                getReentrantReadWriteLockForCell(cell).readLock();
+        lock.lock();
+        try {
+            runTaskOnNodesMaybeFail(quorumNodes, task);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private void runTaskOnNodesMaybeFail(Set<Node> quorumNodes, Consumer<Node> task) {
+        maybeFail();
+        for (Node current : quorumNodes) {
+            task.accept(current);
+            maybeFail();
+        }
+    }
```

Comment:
I think this is fine because `Node` has a `ConcurrentHashMap`. In general, supposing you have a conflict between

1. a PUE and some other operation: if it is a read, they may resolve in either order. If it is a write or PUE, the locks prevent anything.
2. two writes: two racing writes result in last write wins, a racing write and read may or may not have the new value seen
3. two reads: this is fine.

In terms of visibility, any values written will become visible because reads go through a concurrent data structure.